src/sa.py

```diff
@@ -1,15 +1,276 @@
+#    SA = [string[SA[int(i)]:] for i in SA]
+#    SA.sort()
+#    return(SA)
+
+def completely_naive(string):
+    SA = [i for i in range(len(string))]
+    SA = [string[SA[int(i)]:] for i in SA]
+    SA.sort()
+    return(SA)
+
+def test_SA(SA):
+    for i in SA:
+        print(i)
+
+def LCP_naive(SA, string):
+    LCP = [0, 0]
+    for i in range(2, len(SA)):
+        count = 0
+        for j in range(len(string)):
+            if string[SA[i] + j] == string[SA[i - 1] + j]:
+                count += 1
+            else:
+                break
+        LCP.append(count)
+    return LCP
+
+def naive(string):
+    string += "$"
+    SA = completely_naive(string)
+    SA = [len(string) - len(i) for i in SA]
+    LCP = LCP_naive(SA, string)
+    return SA, LCP
+
+def compare_strings(pattern, suffix):
+    loops = min(len(pattern), len(suffix))
+
+    for i in range(loops):
+        if pattern[i] < suffix[i]:
+            return False
+        if pattern[i] > suffix[i]:
+            return True
+
+def SA_search(SA, string, pattern):
+    lo = 0
+    hi = len(SA) - 1
+    while True:
+        mid = int((lo + hi)/2)
+        suffix = string[SA[mid]:]
+        match = compare_strings(pattern, suffix)
+        if match is None:
+            return mid
+        elif hi == lo:
+            return None
+        elif match:
+            lo = mid + 1
+        elif not match:
+            hi = mid - 1
+
+def SA_match(SA, LCP, string, pattern):
+    string += "$"
+    match = SA_search(SA, string, pattern)
+    l = len(pattern)
+    if match is None:
+        return []
+    lo = match
+    hi = match + 1
+
+    while hi < len(LCP):
+        if LCP[hi] >= l:
+            hi += 1
+        else:
+            break
+    while lo > 0:
+        if LCP[lo] >= l:
+            lo -= 1
+        else:
+            break
+    return SA[lo:hi]
+
+def matches_to_SAM(read_file, ref_file):
+    read_name = []
+    reference_name = []
+    match_index = []
+    CIGARS = []
+    match_string = []
+
+    fasta_dict, fastq_dict = fasta_translator(ref_file), fastq_translator(read_file)
+
+    for i in fasta_dict:
+        SA, LCP = naive(fasta_dict[i])
+        for j in fastq_dict:
+            matches_temp = SA_match(SA, LCP, fasta_dict[i], fastq_dict[j])
+            if matches_temp:
+                for match in matches_temp: # Each iteration makes a SAM row
+                    read_name.append(j)
+                    reference_name.append(i)
+                    match_index.append(match + 1)
+                    CIGARS.append(str(len(fastq_dict[j])) + "M")
+                    match_string.append(fastq_dict[j])
+    output = (read_name, reference_name, match_index, CIGARS, match_string)
+    return output
+
+def print_SAM(SAM):
+    for i in range(len(SAM[0])):
+        sys.stdout.write(SAM[0][i] + "\t" + SAM[1][i] + "\t" + str(SAM[2][i]) + "\t" + SAM[3][i] + "\t" + SAM[4][i] + "\n")
+
+SAM = matches_to_SAM(args.reads, args.genome)
+
+print_SAM(SAM)
```

Comment:
Dit problem er at du vil pille i den skabelon jeg har givet jer. Når man går det, så risikerer man at blive straffet. Det er du så blevet.

Du skal have beskyttet den kode du kører når du kører denne fil som et script fra den kode der skal køres når den bliver importeret. Ellers får du fejl når du kører doctest.

Du kan fikse det ved at sætte de sidste to linjer ind under en `if __name__ == '__main__:`

```python
if __name__ == '__main__':
    SAM = matches_to_SAM(args.reads, args.genome)
    print_SAM(SAM)
```